src/main/java/org/folio/helper/BindHelper.java

```diff
@@ -55,48 +65,82 @@ public Future<ReceivingResults> bindPieces(BindPiecesCollection bindPiecesCollec
-    // 1. Get piece records from storage
+    //   1. Get piece records from storage
     return retrievePieceRecords(requestContext)
-      // 2. Update piece isBound flag
+      // 2. Check if there are any outstanding requests for items
+      .compose(piecesGroupedByPoLine -> checkRequestsForPieceItems(piecesGroupedByPoLine, bindPiecesCollection, requestContext))
+      // 3. Update piece isBound flag
       .map(this::updatePieceRecords)
-      // 3. Update currently associated items
-      .map(piecesGroupedByPoLine -> updateItemStatus(piecesGroupedByPoLine, requestContext))
-      // 4. Crate item for pieces with specific fields
+      // 4. Update currently associated items
+      .compose(piecesGroupedByPoLine -> updateItemStatus(piecesGroupedByPoLine, requestContext))
+      // 5. Crate item for pieces with specific fields
       .compose(piecesGroupedByPoLine -> createItemForPiece(piecesGroupedByPoLine, bindPiecesCollection, requestContext))
-      // 5. Update received piece records in the storage
+      // 6. Update received piece records in the storage
       .compose(piecesGroupedByPoLine -> storeUpdatedPieceRecords(piecesGroupedByPoLine, requestContext))
-      // 6. Update Title with new bind items
+      // 7. Update Title with new bind items
       .map(piecesGroupedByPoLine -> updateTitleWithBindItems(piecesGroupedByPoLine, requestContext))
-      // 7. Return results to the client
+      // 8. Return results to the client
       .map(piecesGroupedByPoLine -> prepareResponseBody(piecesGroupedByPoLine, bindPiecesCollection));
   }
 
+  private Future<Map<String, List<Piece>>> checkRequestsForPieceItems(Map<String, List<Piece>> piecesGroupedByPoLine,
+                                                                      BindPiecesCollection bindPiecesCollection,
+                                                                      RequestContext requestContext) {
+    var tenantToItem = mapTenantIdsToItemIds(piecesGroupedByPoLine, requestContext);
+    return GenericCompositeFuture.all(
+        tenantToItem.entrySet().stream()
+          .map(entry -> {
+          var locationContext = RequestContextUtil.createContextWithNewTenantId(requestContext, entry.getKey());
+          return inventoryItemRequestService.getItemsWithActiveRequests(entry.getValue(), locationContext)
+            .compose(items -> {
+              if (!items.isEmpty()) {
+                // requestsAction is required to handle outstanding requests
+                if (Objects.isNull(bindPiecesCollection.getRequestsAction())) {
+                  logger.warn("checkRequestsForPieceItems:: Found outstanding requests on items with ids: {}", items);
+                  throw new HttpException(RestConstants.VALIDATION_ERROR, ErrorCodes.REQUESTS_ACTION_REQUIRED);
+                }
+                else if (bindPiecesCollection.getRequestsAction().equals(TRANSFER)) {
```

Comment:
Why do you need `else if` if you throw an error otherwise?
It's better to handle edge case first:
```
                if (items.isEmpty()) {
                    return Future.succeededFuture();
                }
```
and then continue with general case.
I'd write:
```
              if (!items.isEmpty() && Objects.isNull(bindPiecesCollection.getRequestsAction())) {
                logger.warn("checkRequestsForPieceItems:: Found outstanding requests on items with ids: {}", items);
                throw new HttpException(RestConstants.VALIDATION_ERROR, ErrorCodes.REQUESTS_ACTION_REQUIRED);
              }
              if (!items.isEmpty() && bindPiecesCollection.getRequestsAction().equals(TRANSFER) && !tenantToItem.keySet().isEmpty()) {
                logger.warn("checkRequestsForPieceItems:: All pieces do not have the same receivingTenantId: {}", tenantToItem.keySet());
                throw new HttpException(RestConstants.VALIDATION_ERROR, ErrorCodes.PIECES_HAVE_DIFFERENT_RECEIVING_TENANT_IDS);
              }
              return Future.succeededFuture();
            });
```
This way I see a 2 cases where you throw an error but it's up to you. Main thing is to not keep many lines of code indented as this increases a complexity.

Reply:
Correct, better to get rid of nested conditions.